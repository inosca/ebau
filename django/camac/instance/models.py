import logging

import reversion
from django.conf import settings
from django.db import models

from camac.core.models import HistoryActionConfig
from camac.user.models import User

from ..core import models as core_models

log = logging.getLogger(__name__)


class FormState(models.Model):
    form_state_id = models.AutoField(db_column="FORM_STATE_ID", primary_key=True)
    name = models.CharField(db_column="NAME", max_length=50)

    class Meta:
        managed = True
        db_table = "FORM_STATE"


class Form(core_models.MultilingualModel, models.Model):
    """Represents type of a form."""

    form_id = models.AutoField(db_column="FORM_ID", primary_key=True)
    form_state = models.ForeignKey(
        FormState, models.DO_NOTHING, db_column="FORM_STATE_ID", related_name="+"
    )
    name = models.CharField(db_column="NAME", max_length=500, blank=True, null=True)
    description = models.CharField(
        db_column="DESCRIPTION", max_length=1000, blank=True, null=True
    )
    family = models.ForeignKey(
        "self",
        models.DO_NOTHING,
        db_column="FAMILY",
        related_name="+",
        blank=False,
        null=True,
    )

    class Meta:
        managed = True
        db_table = "FORM"


class FormT(models.Model):
    form = models.ForeignKey(
        Form, models.CASCADE, db_column="FORM_ID", related_name="trans"
    )
    language = models.CharField(db_column="LANGUAGE", max_length=2)
    name = models.CharField(db_column="NAME", max_length=500, blank=True, null=True)
    description = models.CharField(
        db_column="DESCRIPTION", max_length=1000, blank=True, null=True
    )

    class Meta:
        managed = True
        db_table = "FORM_T"


class InstanceState(core_models.MultilingualModel, models.Model):
    instance_state_id = models.AutoField(
        db_column="INSTANCE_STATE_ID", primary_key=True
    )
    name = models.CharField(db_column="NAME", max_length=100, blank=True, null=True)
    sort = models.IntegerField(db_column="SORT", db_index=True, default=0)
    description = models.CharField(
        db_column="DESCRIPTION", max_length=1000, blank=True, null=True
    )

    class Meta:
        managed = True
        db_table = "INSTANCE_STATE"


class InstanceStateT(models.Model):
    instance_state = models.ForeignKey(
        InstanceState,
        models.CASCADE,
        db_column="INSTANCE_STATE_ID",
        related_name="trans",
    )
    language = models.CharField(db_column="LANGUAGE", max_length=2)
    name = models.CharField(db_column="NAME", max_length=100, blank=True, null=True)
    description = models.CharField(
        db_column="DESCRIPTION", max_length=1000, blank=True, null=True
    )

    class Meta:
        managed = True
        db_table = "INSTANCE_STATE_T"


class InstanceStateDescription(models.Model):
    """
    Instance state description.

    Obsolete as integrated into core now. Still added for backwards
    compatability with Kanton URI project.
    """

    instance_state = models.OneToOneField(
        InstanceState,
        models.DO_NOTHING,
        db_column="INSTANCE_STATE_ID",
        primary_key=True,
        related_name="+",
    )
    description = models.CharField(db_column="DESCRIPTION", max_length=255)

    class Meta:
        managed = True
        db_table = "INSTANCE_STATE_DESCRIPTION"


class InstanceGroup(models.Model):
    pass


@reversion.register()
class Instance(models.Model):
    """
    Instance is the case entity of any request.

    Instance is always based on a type of form.
    """

    instance_id = models.AutoField(db_column="INSTANCE_ID", primary_key=True)
    instance_state = models.ForeignKey(
        InstanceState,
        models.DO_NOTHING,
        db_column="INSTANCE_STATE_ID",
        related_name="+",
    )
    form = models.ForeignKey(
        Form, models.DO_NOTHING, db_column="FORM_ID", related_name="+"
    )
    user = models.ForeignKey(
        User, models.DO_NOTHING, db_column="USER_ID", related_name="+"
    )
    group = models.ForeignKey(
        "user.Group", models.DO_NOTHING, db_column="GROUP_ID", related_name="+"
    )
    creation_date = models.DateTimeField(db_column="CREATION_DATE", auto_now_add=True)
    modification_date = models.DateTimeField(
        db_column="MODIFICATION_DATE", auto_now_add=True
    )
    previous_instance_state = models.ForeignKey(
        InstanceState,
        models.DO_NOTHING,
        db_column="PREVIOUS_INSTANCE_STATE_ID",
        related_name="+",
    )
    identifier = models.CharField(
        db_column="IDENTIFIER", max_length=50, blank=True, null=True
    )
    location = models.ForeignKey(
        "user.Location", models.PROTECT, null=True, blank=True, db_column="LOCATION_ID"
    )
    services = models.ManyToManyField("user.Service", through="core.InstanceService")
    case = models.OneToOneField(
        "caluma_workflow.Case",
        null=True,
        blank=True,
        on_delete=models.SET_NULL,
        related_name="instance",
    )
    instance_group = models.ForeignKey(
        InstanceGroup, models.SET_NULL, related_name="instances", null=True
    )

    def _responsible_service_instance_service(self, filter_type=None, **kwargs):
        active_services_settings = settings.APPLICATION.get("ACTIVE_SERVICES", {})

        if filter_type:
            filter_type = filter_type.upper()

            if filter_type not in active_services_settings.keys():  # pragma: no cover
                raise Exception(
                    f"Active service `filter_type` {filter_type} is not configured"
                )

            active_service_config = active_services_settings.get(filter_type)
        else:
            active_service_config = None
            default_active_service_config = None

            for config in active_services_settings.values():
                if config.get("DEFAULT"):
                    default_active_service_config = config

                if any(
                    [
                        self.instance_state.name == current_state
                        and self.previous_instance_state.name == previous_state
                        for current_state, previous_state in config.get(
                            "INSTANCE_STATES", []
                        )
                    ]
                ):
                    active_service_config = config

            active_service_config = (
                active_service_config or default_active_service_config
            )

        service_filters = active_service_config.get("FILTERS", {})

        instance_services = self.instance_services.filter(active=1, **service_filters)
        instance_service = instance_services.order_by("-pk").first()

        if instance_services.count() > 1:
            log.warning(
                f"Instance {self.pk}: Multiple active services, picking most recent one: {instance_service.service.get_name()}!"
            )

        return instance_service.service if instance_service else None

    def responsible_service(self, **kwargs):
        """
        Call application specific method and fallback to active_service.

        Application specific methods have to be named like this:
        _responsible_service_{application_name}
        """
        if settings.APPLICATION.get("USE_INSTANCE_SERVICE"):
            return self._responsible_service_instance_service(**kwargs)

        return self.group.service

    class Meta:
        managed = True
        db_table = "INSTANCE"


class InstanceResponsibility(models.Model):
    instance = models.ForeignKey(
        Instance, models.CASCADE, related_name="responsibilities"
    )
    service = models.ForeignKey(
        "user.Service", models.CASCADE, db_column="SERVICE_ID", related_name="+"
    )
    user = models.ForeignKey(
        User,
        models.DO_NOTHING,
        db_column="USER_ID",
        related_name="responsibilities",
    )

    class Meta:
        unique_together = (("instance", "user", "service"),)


class JournalEntry(models.Model):
    instance = models.ForeignKey(Instance, models.CASCADE, related_name="journal")
    service = models.ForeignKey(
        "user.Service", models.DO_NOTHING, related_name="+", null=True
    )
    user = models.ForeignKey(User, models.DO_NOTHING, related_name="+")
    text = models.TextField(blank=True, null=True)
    creation_date = models.DateTimeField()
    modification_date = models.DateTimeField()
    VISIBILITIES = [
        ("all", "All"),
        ("own_organization", "Own Organization"),
        ("authorities", "Authorities"),
    ]
    visibility = models.CharField(
        max_length=16, choices=VISIBILITIES, default="own_organization"
    )
    duration = models.DurationField(blank=True, null=True)


class HistoryEntry(core_models.MultilingualModel, models.Model):
    instance = models.ForeignKey(Instance, models.CASCADE, related_name="history")
    service = models.ForeignKey(
        "user.Service", models.DO_NOTHING, related_name="+", null=True
    )
    user = models.ForeignKey(User, models.DO_NOTHING, related_name="+")
    created_at = models.DateTimeField(auto_now_add=True)
    title = models.TextField(blank=True, null=True)
    body = models.TextField(blank=True, null=True)
    history_type = models.CharField(
        max_length=20, choices=HistoryActionConfig.HISTORY_TYPES_TUPLE
    )

    def __str__(self):
        return self.get_trans_attr("title")


class HistoryEntryT(models.Model):
    title = models.TextField()
    body = models.TextField(blank=True, null=True)
    history_entry = models.ForeignKey(
        HistoryEntry, models.CASCADE, related_name="trans"
    )
    language = models.CharField(max_length=2)


class Issue(models.Model):
    STATE_OPEN = "open"
    STATE_DELAYED = "delayed"
    STATE_DONE = "done"
    STATE_CHOICES = (STATE_OPEN, STATE_DELAYED, STATE_DONE)
    STATE_CHOICES_TUPLE = ((choice, choice) for choice in STATE_CHOICES)

    instance = models.ForeignKey(Instance, models.CASCADE, related_name="issues")
    group = models.ForeignKey("user.Group", models.DO_NOTHING, related_name="+")
    service = models.ForeignKey(
        "user.Service", models.DO_NOTHING, related_name="+", null=True
    )
    user = models.ForeignKey(
        User, models.DO_NOTHING, related_name="+", blank=True, null=True
    )
    deadline_date = models.DateField()
    state = models.CharField(
        max_length=20, choices=STATE_CHOICES_TUPLE, default=STATE_OPEN
    )
    text = models.TextField()


class IssueTemplate(models.Model):
    group = models.ForeignKey("user.Group", models.DO_NOTHING, related_name="+")
    service = models.ForeignKey(
        "user.Service", models.DO_NOTHING, related_name="+", null=True
    )
    user = models.ForeignKey(
        User, models.DO_NOTHING, related_name="+", blank=True, null=True
    )
    deadline_length = models.PositiveIntegerField()
    text = models.TextField()


class IssueTemplateSet(models.Model):
    group = models.ForeignKey("user.Group", models.DO_NOTHING, related_name="+")
    service = models.ForeignKey(
        "user.Service", models.DO_NOTHING, related_name="+", null=True
    )
    issue_templates = models.ManyToManyField(
        IssueTemplate, related_name="issue_template_sets", blank=True
    )
    name = models.CharField(max_length=500)


@reversion.register()
class FormField(models.Model):
    """
    Represents fields of an instance form.

    Which form type the field references is assigned on instance itself.
    """

    instance = models.ForeignKey(Instance, models.CASCADE, related_name="fields")
    name = models.CharField(max_length=500)
    value = models.JSONField()

    class Meta:
        unique_together = (("instance", "name"),)
