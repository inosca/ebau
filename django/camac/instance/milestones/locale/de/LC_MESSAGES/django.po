# SOME DESCRIPTIVE TITLE.
# Copyright (C) YEAR THE PACKAGE'S COPYRIGHT HOLDER
# This file is distributed under the same license as the PACKAGE package.
# FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
#
msgid ""
msgstr ""
"Project-Id-Version: \n"
"Report-Msgid-Bugs-To: \n"
"POT-Creation-Date: 2024-09-11 17:25+0200\n"
"PO-Revision-Date: 2024-08-29 13:33+0200\n"
"Last-Translator: \n"
"Language-Team: \n"
"Language: de\n"
"MIME-Version: 1.0\n"
"Content-Type: text/plain; charset=UTF-8\n"
"Content-Transfer-Encoding: 8bit\n"
"Plural-Forms: nplurals=2; plural=(n != 1);\n"
"X-Generator: Poedit 3.0.1\n"

#: camac/instance/milestones/serializers.py
msgid "regular process"
msgstr "Ordentliches Verfahren"

#: camac/instance/milestones/serializers.py
msgid "instance submitted"
msgstr "Dossier erfasst"

#: camac/instance/milestones/serializers.py
msgid "Paper submission"
msgstr "Eingang Post"

#: camac/instance/milestones/serializers.py
msgid "Additional demand"
msgstr "Nachforderung Gemeindebaubehörde"

#: camac/instance/milestones/serializers.py
msgid "Check completed"
msgstr "Vollständigkeitsprüfung abgeschlossen"

#: camac/instance/milestones/serializers.py
msgid "Instance complete"
msgstr "Dossier vollständig"

#: camac/instance/milestones/serializers.py
msgid "Review building commission"
msgstr "Prüfung durch Gemeinde"

#: camac/instance/milestones/serializers.py
msgid "Puclication date"
msgstr "Publikationsdatum"

#: camac/instance/milestones/serializers.py
msgid "Receipt of objections"
msgstr "Eingang von Einsprachen"

#: camac/instance/milestones/serializers.py
msgid "Forwarding to KOOR"
msgstr "Weiterleitung an KOOR"

#: camac/instance/milestones/serializers.py
msgid "Start circulation"
msgstr "Start Zirkulation"

#: camac/instance/milestones/serializers.py
msgid "Distribution completed"
msgstr "Zirkulation beendet"

#: camac/instance/milestones/serializers.py
msgid "Statement - preliminary decision"
msgstr "Stellungnahme - Vorentscheid"

#: camac/instance/milestones/serializers.py
msgid "Building decision"
msgstr "Bau- und Einspracheentscheid"

#: camac/instance/milestones/serializers.py
msgid "notice to geometer"
msgstr "Meldung Baubewilligung an Geometer"

#: camac/instance/milestones/serializers.py
msgid "Receipt confirmation of decision documents"
msgstr "Empfangsbestätigung Entscheiddokumente"

#: camac/instance/milestones/serializers.py
msgid "Receipt confirmation of preliminary decision"
msgstr "Empfangsbestätigung Vorentscheid"

#: camac/instance/milestones/serializers.py
msgid "notice to gebaeudeschaetzung"
msgstr "Meldung Baubewilligung an Gebäudeschätzung"

#: camac/instance/milestones/serializers.py
msgid "Building permit valid until"
msgstr "Baubewilligung gültig bis"
