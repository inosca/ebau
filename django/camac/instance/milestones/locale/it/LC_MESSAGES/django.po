# SOME DESCRIPTIVE TITLE.
# Copyright (C) YEAR THE PACKAGE'S COPYRIGHT HOLDER
# This file is distributed under the same license as the PACKAGE package.
# FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
#
#, fuzzy
msgid ""
msgstr ""
"Project-Id-Version: PACKAGE VERSION\n"
"Report-Msgid-Bugs-To: \n"
"POT-Creation-Date: 2024-09-11 17:25+0200\n"
"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\n"
"Last-Translator: FULL NAME <EMAIL@ADDRESS>\n"
"Language-Team: LANGUAGE <LL@li.org>\n"
"Language: \n"
"MIME-Version: 1.0\n"
"Content-Type: text/plain; charset=UTF-8\n"
"Content-Transfer-Encoding: 8bit\n"
"Plural-Forms: nplurals=2; plural=(n != 1);\n"

#: camac/instance/milestones/serializers.py
msgid "regular process"
msgstr ""

#: camac/instance/milestones/serializers.py
msgid "instance submitted"
msgstr ""

#: camac/instance/milestones/serializers.py
msgid "Paper submission"
msgstr ""

#: camac/instance/milestones/serializers.py
msgid "Additional demand"
msgstr ""

#: camac/instance/milestones/serializers.py
msgid "Check completed"
msgstr ""

#: camac/instance/milestones/serializers.py
msgid "Instance complete"
msgstr ""

#: camac/instance/milestones/serializers.py
msgid "Review building commission"
msgstr ""

#: camac/instance/milestones/serializers.py
msgid "Publication date"
msgstr ""

#: camac/instance/milestones/serializers.py
msgid "Receipt of objections"
msgstr ""

#: camac/instance/milestones/serializers.py
msgid "Forwarding to KOOR"
msgstr ""

#: camac/instance/milestones/serializers.py
msgid "Start circulation"
msgstr ""

#: camac/instance/milestones/serializers.py
msgid "Distribution completed"
msgstr ""

#: camac/instance/milestones/serializers.py
msgid "Statement - preliminary decision"
msgstr ""

#: camac/instance/milestones/serializers.py
msgid "Building decision"
msgstr ""

#: camac/instance/milestones/serializers.py
msgid "notice to geometer"
msgstr ""

#: camac/instance/milestones/serializers.py
msgid "Receipt confirmation of decision documents"
msgstr ""

#: camac/instance/milestones/serializers.py
msgid "Receipt confirmation of preliminary decision"
msgstr ""

#: camac/instance/milestones/serializers.py
msgid "notice to gebaeudeschaetzung"
msgstr ""

#: camac/instance/milestones/serializers.py
msgid "Building permit valid until"
msgstr ""
