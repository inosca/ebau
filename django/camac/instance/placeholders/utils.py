from datetime import date, datetime
from typing import Any, Optional, Union

from babel.dates import format_date
from django.utils.translation import get_language


def clean_join(*parts: Any, separator: Optional[str] = " ") -> str:
    """Join all truthy arguments as trimmed string with a given separator.

    >>> clean_join(" John", None, "", " Smith ")
    "John Smith"
    """

    return separator.join([str(part).strip() for part in parts if part]).strip()


def get_option_label(option: Union[dict, None]) -> Union[str, None]:
    """Get the label of an option.

    >>> get_option_label({"slug": "test-option", "label": "Test Option"})
    "Test Option"
    """
    return option.get("label") if option else None


def get_person_address_1(person: dict) -> str:
    """Extract the first address line from a person dictionary.

    >>> get_person_address_1({
        "street_name": "Test Avenue",
        "street_number": 10,
    })
    "Test Avenue 10"
    """

    return clean_join(person.get("street"), person.get("street_number"))


def get_person_address_2(person):
    """Extract the second address line from a person dictionary.

    >>> get_person_address_1({
        "zip": 9999,
        "city": "Testcity"
    })
    "9999 Testcity"
    """

    return clean_join(person.get("zip"), person.get("town"))


def get_tel_and_email(person):
    return clean_join(person.get("tel"), person.get("email"), separator=", ")


def get_person_name(
    person: dict,
    include_name: bool = True,
    include_juristic_name: bool = True,
) -> str:
    """Extract the name from a person dictionary.

    >>> get_person_name({
        "first_name": "John",
        "last_name": "Smith",
        "is_juristic_person": True,
        "juristic_person_name": "ACME Inc.",
    })
    "ACME Inc., John Smith"
    """

    parts = []

    if include_juristic_name and person.get("is_juristic_person"):
        parts.append(clean_join(person.get("juristic_name")))

    if include_name:
        parts.append(clean_join(person.get("first_name"), person.get("last_name")))

    return clean_join(*parts, separator=", ")


def human_readable_date(value: Union[datetime, date, None]) -> str:
    """Format date or datetime to a human readable date.

    >>> human_readable_date(date(2021, 10, 4))
    "4. October 2021"
    """

    if not value:
        return None

    return format_date(value, "long", locale=get_language())
