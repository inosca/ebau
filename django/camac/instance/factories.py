from random import randrange

import pytz
from factory import Faker, SubFactory, fuzzy, post_generation
from factory.django import DjangoModelFactory

from camac.applicants.factories import ApplicantFactory
from camac.core.models import HistoryActionConfig
from camac.user.factories import (
    GroupFactory,
    LocationFactory,
    ServiceFactory,
    UserFactory,
)

from . import models


class FormStateFactory(DjangoModelFactory):
    # Set to Published because it is the default FormState
    # other FormStates are usually not accessible outside from the admin
    name = "Published"

    class Meta:
        model = models.FormState


class FormFactory(DjangoModelFactory):
    name = Faker("name")
    description = Faker("name")
    form_state = SubFactory(FormStateFactory)
    family = None

    class Meta:
        model = models.Form


class InstanceStateFactory(DjangoModelFactory):
    name = Faker("name")
    sort = 0

    class Meta:
        model = models.InstanceState


class InstanceGroupFactory(DjangoModelFactory):
    class Meta:
        model = models.InstanceGroup


class InstanceFactory(DjangoModelFactory):
    identifier = None
    instance_state = SubFactory(InstanceStateFactory)
    previous_instance_state = SubFactory(InstanceStateFactory)
    form = SubFactory(FormFactory)
    user = SubFactory(UserFactory)
    group = SubFactory(GroupFactory)
    location = SubFactory(LocationFactory)
    creation_date = Faker("past_datetime", tzinfo=pytz.UTC)
    modification_date = Faker("past_datetime", tzinfo=pytz.UTC)

    @post_generation
    def involved_applicants(self, create, extracted):
        return ApplicantFactory.create_batch(
            1, instance=self, user=self.user, invitee=self.user
        )

    class Meta:
        model = models.Instance


class FormFieldFactory(DjangoModelFactory):
    instance = SubFactory(InstanceFactory)
    name = Faker("name")
    value = Faker("name")

    class Meta:
        model = models.FormField


class InstanceResponsibilityFactory(DjangoModelFactory):
    service = SubFactory(ServiceFactory)
    user = SubFactory(UserFactory)
    instance = SubFactory(InstanceFactory)

    class Meta:
        model = models.InstanceResponsibility


class JournalEntryFactory(DjangoModelFactory):
    instance = SubFactory(InstanceFactory)
    user = SubFactory(UserFactory)
    service = SubFactory(ServiceFactory)
    creation_date = Faker("past_datetime", tzinfo=pytz.UTC)
    modification_date = Faker("past_datetime", tzinfo=pytz.UTC)
    text = Faker("text")
    visibility = "own_organization"
    duration = Faker("time_delta")

    class Meta:
        model = models.JournalEntry


class HistoryEntryFactory(DjangoModelFactory):
    instance = SubFactory(InstanceFactory)
    user = SubFactory(UserFactory)
    service = SubFactory(ServiceFactory)
    created_at = Faker("past_datetime", tzinfo=pytz.UTC)
    title = Faker("text")
    body = Faker("text")
    history_type = fuzzy.FuzzyChoice(HistoryActionConfig.HISTORY_TYPES)

    class Meta:
        model = models.HistoryEntry


class HistoryEntryTFactory(DjangoModelFactory):
    title = Faker("text")
    body = Faker("text")
    history_entry = SubFactory(HistoryEntryFactory)
    language = "de"

    class Meta:
        model = models.HistoryEntryT


class IssueFactory(DjangoModelFactory):
    instance = SubFactory(InstanceFactory)
    user = SubFactory(UserFactory)
    group = SubFactory(GroupFactory)
    service = SubFactory(ServiceFactory)
    deadline_date = Faker("future_date")
    text = Faker("text")
    state = models.Issue.STATE_OPEN

    class Meta:
        model = models.Issue


class IssueTemplateFactory(DjangoModelFactory):
    user = SubFactory(UserFactory)
    group = SubFactory(GroupFactory)
    service = SubFactory(ServiceFactory)
    deadline_length = randrange(1, 10)
    text = Faker("text")

    class Meta:
        model = models.IssueTemplate


class IssueTemplateSetFactory(DjangoModelFactory):
    group = SubFactory(GroupFactory)
    service = SubFactory(ServiceFactory)
    name = Faker("sentence")

    class Meta:
        model = models.IssueTemplateSet


class IssueTemplateSetIssueTemplateFactory(DjangoModelFactory):
    issuetemplate = SubFactory(IssueTemplateFactory)
    issuetemplateset = SubFactory(IssueTemplateSetFactory)

    class Meta:
        model = models.IssueTemplateSet.issue_templates.through
