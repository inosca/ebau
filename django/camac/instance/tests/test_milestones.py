from datetime import datetime

import pytest
from caluma.caluma_workflow.factories import WorkItemFactory
from caluma.caluma_workflow.models import WorkItem
from django.urls import reverse
from django.utils.timezone import make_aware
from rest_framework import status

from camac.constants import kt_uri as uri_constants


@pytest.mark.parametrize(
    "role__name",
    [
        ("Municipality"),
    ],
)
@pytest.mark.parametrize("is_paper_answer", ["is-paper-no", "is-paper-yes"])
@pytest.mark.freeze_time("2023-12-31")
@pytest.mark.django_db(
    transaction=True, reset_sequences=True
)  # always reset instance id
def test_milestones_ur(
    db,
    role,
    admin_client,
    ur_instance,
    settings,
    application_settings,
    snapshot,
    is_paper_answer,
    answer_factory,
):
    application_settings["CALUMA"]["SUBMIT_TASKS"] = ["submit"]
    application_settings["COORDINATION_SERVICE_IDS"] = [
        uri_constants.KOOR_BG_SERVICE_ID
    ]
    settings.APPLICATION_NAME = "kt_uri"

    answer_factory(
        document=ur_instance.case.document,
        question_id="is-paper",
        value=is_paper_answer,
    )

    submit_task = ur_instance.case.work_items.get(task_id="submit")
    submit_task.closed_at = make_aware(datetime(2023, 1, 1, 20, 0, 0))
    submit_task.save()

    for task_slug in ["check-permit", "decision"]:
        WorkItemFactory(
            task_id=task_slug,
            case=ur_instance.case,
            closed_at=make_aware(datetime(2023, 1, 1, 20, 0, 0)),
            status=WorkItem.STATUS_COMPLETED,
        )

    # top level circulation (Gemeinde -> KOOR)
    WorkItemFactory(
        task_id="inquiry",
        case=ur_instance.case,
        created_at=make_aware(datetime(2023, 1, 1, 20, 0, 0)),
        addressed_groups=[uri_constants.KOOR_BG_SERVICE_ID],
        controlling_groups=[],
        status=WorkItem.STATUS_COMPLETED,
        closed_at=make_aware(datetime(2023, 1, 1, 20, 0, 0)),
    )

    # Nested KOOR circulation
    WorkItemFactory(
        task_id="inquiry",
        case=ur_instance.case,
        created_at=make_aware(datetime(2023, 1, 1, 20, 0, 0)),
        controlling_groups=[uri_constants.KOOR_BG_SERVICE_ID],
        status=WorkItem.STATUS_COMPLETED,
        closed_at=make_aware(datetime(2023, 1, 1, 20, 0, 0)),
    )

    url = reverse("instance-milestones", args=[ur_instance.pk])
    response = admin_client.get(url)

    assert response.status_code == status.HTTP_200_OK
    snapshot.assert_match(response.json())
