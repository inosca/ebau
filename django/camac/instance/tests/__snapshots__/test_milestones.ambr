# serializer version: 1
# name: test_milestones_ur[is-paper-no-complete-check-vollstaendigkeitspruefung-complete-False-Sekretariat der Gemeindebaubeh\xf6rde]
  dict({
    'sections': list([
      dict({
        'fields': list([
          dict({
            'label': 'Dossier erfasst',
            'slug': 'instance-submitted',
            'value': '2023-01-01T19:00:00Z',
          }),
          dict({
            'label': 'Eingang Post',
            'slug': 'paper-submission',
            'value': list([
            ]),
          }),
          dict({
            'label': 'Nachforderung Gemeindebaubehörde',
            'slug': 'additional-demand',
            'value': list([
            ]),
          }),
          dict({
            'label': 'Dossier vollstandig',
            'slug': 'check-completed',
            'value': list([
              '2023-01-01T19:00:00Z',
            ]),
          }),
          dict({
            'label': 'Dossier vollstandig',
            'slug': 'instance-complete',
            'value': '2023-01-01T19:00:00Z',
          }),
          dict({
            'label': 'Prüfung durch Gemeinde',
            'slug': 'review-building-commission',
            'value': list([
            ]),
          }),
          dict({
            'label': 'Start Zirkulation',
            'slug': 'start-circulation',
            'value': list([
            ]),
          }),
          dict({
            'label': 'Zirkulation beendet',
            'slug': 'distribution-completed',
            'value': list([
            ]),
          }),
          dict({
            'label': 'Stellungnahme - Vorentscheid',
            'slug': 'decision',
            'value': list([
              '2023-01-01T19:00:00Z',
            ]),
          }),
          dict({
            'label': 'Meldung Baubewilligung an Geometer',
            'slug': 'notice-to-geometer',
            'value': list([
            ]),
          }),
          dict({
            'label': 'Meldung Baubewilligung an Gebäudeschätzung',
            'slug': 'notice-to-gebaeudeschaetzung',
            'value': list([
            ]),
          }),
          dict({
            'label': 'Baubewilligung gültig bis',
            'slug': 'building-permit-valid-until',
            'value': '2024-01-01T19:00:00Z',
          }),
        ]),
        'label': 'Ordentliches Verfahren',
        'slug': 'regular-process',
      }),
    ]),
  })
# ---
# name: test_milestones_ur[is-paper-no-complete-check-vollstaendigkeitspruefung-incomplete-False-Sekretariat der Gemeindebaubeh\xf6rde]
  dict({
    'sections': list([
      dict({
        'fields': list([
          dict({
            'label': 'Dossier erfasst',
            'slug': 'instance-submitted',
            'value': '2023-01-01T19:00:00Z',
          }),
          dict({
            'label': 'Eingang Post',
            'slug': 'paper-submission',
            'value': list([
            ]),
          }),
          dict({
            'label': 'Nachforderung Gemeindebaubehörde',
            'slug': 'additional-demand',
            'value': list([
            ]),
          }),
          dict({
            'label': 'Dossier vollstandig',
            'slug': 'check-completed',
            'value': list([
              '2023-01-01T19:00:00Z',
            ]),
          }),
          dict({
            'label': 'Dossier vollstandig',
            'slug': 'instance-complete',
            'value': '2023-01-01T19:00:00Z',
          }),
          dict({
            'label': 'Prüfung durch Gemeinde',
            'slug': 'review-building-commission',
            'value': list([
            ]),
          }),
          dict({
            'label': 'Start Zirkulation',
            'slug': 'start-circulation',
            'value': list([
            ]),
          }),
          dict({
            'label': 'Zirkulation beendet',
            'slug': 'distribution-completed',
            'value': list([
            ]),
          }),
          dict({
            'label': 'Stellungnahme - Vorentscheid',
            'slug': 'decision',
            'value': list([
              '2023-01-01T19:00:00Z',
            ]),
          }),
          dict({
            'label': 'Meldung Baubewilligung an Geometer',
            'slug': 'notice-to-geometer',
            'value': list([
            ]),
          }),
          dict({
            'label': 'Meldung Baubewilligung an Gebäudeschätzung',
            'slug': 'notice-to-gebaeudeschaetzung',
            'value': list([
            ]),
          }),
          dict({
            'label': 'Baubewilligung gültig bis',
            'slug': 'building-permit-valid-until',
            'value': '2024-01-01T19:00:00Z',
          }),
        ]),
        'label': 'Ordentliches Verfahren',
        'slug': 'regular-process',
      }),
    ]),
  })
# ---
# name: test_milestones_ur[is-paper-no-complete-check-vollstaendigkeitspruefung-incomplete-True-Sekretariat der Gemeindebaubeh\xf6rde]
  dict({
    'sections': list([
      dict({
        'fields': list([
          dict({
            'label': 'Dossier erfasst',
            'slug': 'instance-submitted',
            'value': '2023-01-01T19:00:00Z',
          }),
          dict({
            'label': 'Eingang Post',
            'slug': 'paper-submission',
            'value': list([
            ]),
          }),
          dict({
            'label': 'Nachforderung Gemeindebaubehörde',
            'slug': 'additional-demand',
            'value': list([
            ]),
          }),
          dict({
            'label': 'Dossier vollstandig',
            'slug': 'check-completed',
            'value': list([
              '2023-01-01T19:00:00Z',
            ]),
          }),
          dict({
            'label': 'Dossier vollstandig',
            'slug': 'instance-complete',
            'value': None,
          }),
          dict({
            'label': 'Prüfung durch Gemeinde',
            'slug': 'review-building-commission',
            'value': list([
            ]),
          }),
          dict({
            'label': 'Start Zirkulation',
            'slug': 'start-circulation',
            'value': list([
            ]),
          }),
          dict({
            'label': 'Zirkulation beendet',
            'slug': 'distribution-completed',
            'value': list([
            ]),
          }),
          dict({
            'label': 'Stellungnahme - Vorentscheid',
            'slug': 'decision',
            'value': list([
              '2023-01-01T19:00:00Z',
            ]),
          }),
          dict({
            'label': 'Meldung Baubewilligung an Geometer',
            'slug': 'notice-to-geometer',
            'value': list([
            ]),
          }),
          dict({
            'label': 'Meldung Baubewilligung an Gebäudeschätzung',
            'slug': 'notice-to-gebaeudeschaetzung',
            'value': list([
            ]),
          }),
          dict({
            'label': 'Baubewilligung gültig bis',
            'slug': 'building-permit-valid-until',
            'value': '2024-01-01T19:00:00Z',
          }),
        ]),
        'label': 'Ordentliches Verfahren',
        'slug': 'regular-process',
      }),
    ]),
  })
# ---
# name: test_milestones_ur[is-paper-yes-complete-check-vollstaendigkeitspruefung-complete-False-Sekretariat der Gemeindebaubeh\xf6rde]
  dict({
    'sections': list([
      dict({
        'fields': list([
          dict({
            'label': 'Dossier erfasst',
            'slug': 'instance-submitted',
            'value': list([
            ]),
          }),
          dict({
            'label': 'Eingang Post',
            'slug': 'paper-submission',
            'value': '2023-01-01T19:00:00Z',
          }),
          dict({
            'label': 'Nachforderung Gemeindebaubehörde',
            'slug': 'additional-demand',
            'value': list([
            ]),
          }),
          dict({
            'label': 'Dossier vollstandig',
            'slug': 'check-completed',
            'value': list([
              '2023-01-01T19:00:00Z',
            ]),
          }),
          dict({
            'label': 'Dossier vollstandig',
            'slug': 'instance-complete',
            'value': '2023-01-01T19:00:00Z',
          }),
          dict({
            'label': 'Prüfung durch Gemeinde',
            'slug': 'review-building-commission',
            'value': list([
            ]),
          }),
          dict({
            'label': 'Start Zirkulation',
            'slug': 'start-circulation',
            'value': list([
            ]),
          }),
          dict({
            'label': 'Zirkulation beendet',
            'slug': 'distribution-completed',
            'value': list([
            ]),
          }),
          dict({
            'label': 'Stellungnahme - Vorentscheid',
            'slug': 'decision',
            'value': list([
              '2023-01-01T19:00:00Z',
            ]),
          }),
          dict({
            'label': 'Meldung Baubewilligung an Geometer',
            'slug': 'notice-to-geometer',
            'value': list([
            ]),
          }),
          dict({
            'label': 'Meldung Baubewilligung an Gebäudeschätzung',
            'slug': 'notice-to-gebaeudeschaetzung',
            'value': list([
            ]),
          }),
          dict({
            'label': 'Baubewilligung gültig bis',
            'slug': 'building-permit-valid-until',
            'value': '2024-01-01T19:00:00Z',
          }),
        ]),
        'label': 'Ordentliches Verfahren',
        'slug': 'regular-process',
      }),
    ]),
  })
# ---
